google/cloud/aiplatform/_pipeline_based_service/pipeline_based_service.py

```diff
@@ -0,0 +1,300 @@
+        gca_pipeline_state.PipelineState.PIPELINE_STATE_PAUSED,
+    ]
+)
+
+
+class _VertexAiPipelineBasedService(base.VertexAiStatefulResource):
+    """Base class for Vertex AI Pipeline based services."""
+
+    client_class = utils.PipelineJobClientWithOverride
+    _resource_noun = "pipelineJob"
+    _delete_method = "delete_pipeline_job"
+    _getter_method = "get_pipeline_job"
+    _list_method = "list_pipeline_jobs"
+    _parse_resource_name_method = "parse_pipeline_job_path"
+    _format_resource_name_method = "pipeline_job_path"
+
+    _valid_done_states = _PIPELINE_COMPLETE_STATES
+
+    @property
+    @classmethod
+    @abc.abstractmethod
+    def _template_ref(self) -> Dict[str, str]:
+        """A dictionary of the pipeline template URLs for this servicewhere the key is
+        an identifier for that template and the value is the url of that pipeline template.
+
+        For example: {"tabular_classification": "gs://path/to/tabular/pipeline/template.json"}
+
+        """
+        pass
+
+    @property
+    @abc.abstractmethod
+    def _metadata_output_artifact(self) -> Optional[str]:
+        """The ML Metadata output artifact resource URI from the completed pipeline run."""
+        pass
+
+    @property
+    def backing_pipeline_job(self) -> "pipeline_jobs.PipelineJob":
+        """The PipelineJob associated with the resource."""
+        return pipeline_jobs.PipelineJob.get(resource_name=self.resource_name)
+
+    @property
+    def pipeline_console_uri(self) -> str:
+        """The console URI of the PipelineJob created by the service."""
+        if self.backing_pipeline_job:
+            return self.backing_pipeline_job._dashboard_uri()
+
+    @property
+    def state(self) -> Optional[str]:
+        """The state of the Pipeline run associated with the service."""
+        if self.backing_pipeline_job:
+            return self.backing_pipeline_job.state
+        return None
+
+    def _validate_pipeline_template_matches_service(
+        self, pipeline_job: "pipeline_jobs.PipelineJob"
+    ):
+        """Utility function to validate that the passed in pipeline ID matches
+        the template of the Pipeline Based Service.
+        Raises:
+            ValueError: if the provided pipeline ID doesn't match the pipeline service.
+        """
+
+        pipeline_match = False
+
+        for pipeline_template in self._template_ref.values():
+            # TODO: should this validate the whole pipelineSpec or just the components level?
+            service_pipeline_json = yaml_utils.load_yaml(pipeline_template)[
+                "pipelineSpec"
+            ]["components"]
+            current_pipeline_json = pipeline_job.to_dict()["pipelineSpec"]["components"]
+
+            if service_pipeline_json == current_pipeline_json:
+                pipeline_match = True
+
+        if not pipeline_match:
+            raise ValueError(
+                f"The provided pipeline template is not compatible with {self.__class__.__name__}"
```

Comment:
Since `_template_ref` is a private property, should we still log them? It would probably be useful to anyone implementing PipelineBasedService, but I wonder if it would confuse end users.